contracts/button/button.rs

```diff
@@ -0,0 +1,399 @@
+    pub presses: Mapping<AccountId, u32>,
+    /// stores keys to `presses` because Mapping is not an Iterator. Heap-allocated so we might need Map<u32, AccountId> if it grows out of proportion
+    pub press_accounts: Vec<AccountId>,
+    /// stores total sum of user scores
+    pub total_scores: u32,
+    /// stores the last account that pressed The Button
+    pub last_presser: Option<AccountId>,
+    /// block number of the last press
+    pub last_press: u32,
+    /// AccountId of the ERC20 ButtonToken instance on-chain
+    pub button_token: AccountId,
+    /// accounts whitelisted to play the game
+    pub can_play: Mapping<AccountId, bool>,
+    /// access control contract
+    pub access_control: AccountId,
+}
+
+/// Concrete implementations of the games API
+///
+/// Implementing contract needs to return ButtonData read-only and mutably
+/// as well as implement `score` the logic that based on the current block number and the contract storage state returns a users score.
+/// Remaining methods have default implementations that can be overriden as needed.
+///
+/// NOTE: no contract events are being emitted, so the implementing contract is responsible for defining and emitting those.
+pub trait ButtonGame {
+    /// Getter for the button data
+    fn get(&self) -> &ButtonData;
+
+    /// Mutable getter for the button data
+    fn get_mut(&mut self) -> &mut ButtonData;
+
+    /// Logic for calculating user score given the particular games rules
+    fn score(&self, now: u32) -> u32;
+
+    fn is_dead(&self, now: u32) -> bool {
+        now > self.deadline()
+    }
+
+    fn deadline(&self) -> u32 {
+        let ButtonData {
+            last_press,
+            button_lifetime,
+            ..
+        } = self.get();
+        last_press + button_lifetime
+    }
+
+    fn score_of(&self, user: AccountId) -> u32 {
+        self.get().presses.get(&user).unwrap_or(0)
+    }
+
+    fn can_play(&self, user: AccountId) -> bool {
+        self.get().can_play.get(&user).unwrap_or(false)
+    }
+
+    fn access_control(&self) -> AccountId {
+        self.get().access_control
+    }
+
+    fn set_access_control(
+        &mut self,
+        new_access_control: AccountId,
+        caller: AccountId,
+        this: AccountId,
+    ) -> Result<()>
+    where
+        Self: AccessControlled,
+    {
+        let required_role = Role::Owner(this);
+        self.check_role(caller, required_role)?;
+        self.get_mut().access_control = new_access_control;
+        Ok(())
+    }
+
+    fn last_presser(&self) -> Option<AccountId> {
+        self.get().last_presser
+    }
+
+    fn button_token(&self) -> Result<AccountId> {
+        Ok(self.get().button_token)
+    }
+
+    fn balance(&self, balance_of_selector: [u8; 4], this: AccountId) -> Result<Balance> {
+        let button_token = self.get().button_token;
+        let balance = build_call::<DefaultEnvironment>()
```

Comment:
I'm not familiar with environments here, but maybe we should be generic over it?

Reply:
sure, why not